crates/shared/src/request_id.rs

```diff
@@ -0,0 +1,84 @@
+//! This module supplies the tools to associate 1 identifier with a task.
+//! That identifier is accessable globally just for that task. The idea
+//! is that this identifier is supposed to tie together related logs. That
+//! is easy to accomplish in a single process (simply use a tracing span)
+//! but if you want to tie together logs across multiple processes things
+//! can get messier.
+//! The most obvious option is to take that identifier and pass that through
+//! the process until you make some request to another process and give that
+//! process the identifier in your request.
+//! However, if would do that the identifier would basically show up everywhere
+//! although other components don't care about it and it doesn't even change
+//! any behaviour in the process.
+//! Instead we use task local storage that is globally visible but only
+//! individual to each task. That way we can populate the storage with the
+//! identifier once and not care about dragging it through the code base.
+//! And when we issue requests to another process we can simply fetch the
+//! current identifier specific to our task and send that along with the
+//! request.
+tokio::task_local! {
+    pub static REQUEST_ID: std::cell::RefCell<String>;
+}
+
+/// Tries to read the `request_id` from this task's storage.
+/// Returns `None` if task local storage was not initialized or is empty.
+pub fn get_task_local_storage() -> Option<String> {
+    let mut id = None;
+    let _ = REQUEST_ID.try_with(|cell| {
+        id = Some(cell.borrow().clone());
+    });
+    id
+}
+
+/// Tries to write the `request_id` to the task local storage.
+/// Fails if this task has no task local storage.
+pub fn set_task_local_storage(request_id: String) {
+    let _ = REQUEST_ID.try_with(|storage| {
+        *storage.borrow_mut() = request_id;
+    });
+}
+
+/// Takes a `tower::Service` and embeds it in a `make_service` function that
+/// spawns one of these services per incoming request.
+/// But crucially before spawning that service task local storage will be
+/// initialized with some request id.
+/// Either that gets taken from the requests `X-REQUEST-ID` header of if that's
+/// missing a globally unique request number will be generated.
+#[macro_export]
+macro_rules! make_service_with_task_local_storage {
```

Comment:
Is there an advantage of having this be a macro vs. a templated function (I personally find macros harder to understand due to missing type annotations)?

Reply:
I would also have preferred a templated function but I was not able to get it working. 🙈 
The `tower` traits and how `axum` and `hyper` are using them is still a mystery to me.
But if we just put the code in a macro the compiler is able to figure things out on its own.